File: components/client-disp-dev/client-disp-dev/src/shared/lib/msw/handlers/vehicles.ts
import { delay, http, HttpResponse } from 'msw';

import type { Vehicle, VehiclesResponse } from '@/shared/api/endpoints/vehicles';

import { mswConfig } from '../config';
import { generateVehicles } from '../generators';

const vehiclesDb: Vehicle[] = generateVehicles(mswConfig.paginatedListSize);

export const vehiclesHandlers = [
  // GET /api/vehicles
  http.get('/api/vehicles', async ({ request }) => {
    await delay(mswConfig.delay);

    const url = new URL(request.url);
    const page = Number(url.searchParams.get('page')) || 1;
    const size = Number(url.searchParams.get('size')) || 100;
    const vehicleType = url.searchParams.get('vehicle_type');

    let filtered = vehiclesDb;

    if (vehicleType) {
      filtered = filtered.filter((v) => v.vehicle_type === vehicleType);
    }

    const total = filtered.length;
    const pages = Math.ceil(total / size);
    const start = (page - 1) * size;
    const items = filtered.slice(start, start + size);

    const response: VehiclesResponse = {
      items,
      total,
      page,
      size,
      pages,
    };

    return HttpResponse.json(response);
  }),

  // GET /api/vehicles/:id
  http.get('/api/vehicles/:id', async ({ params }) => {
    await delay(mswConfig.delay);

    const id = Number(params.id);
    const vehicle = vehiclesDb.find((v) => v.id === id);

    if (!vehicle) {
      return new HttpResponse(null, { status: 404 });
    }

    return HttpResponse.json(vehicle);
  }),

  // POST /api/vehicles
  http.post('/api/vehicles', async ({ request }) => {
    await delay(mswConfig.delay);

    const body = (await request.json()) as Partial<Vehicle>;
    const newId = Math.max(...vehiclesDb.map((v) => v.id), 0) + 1;
    const now = new Date().toISOString();

    const newVehicle: Vehicle = {
      id: newId,
      enterprise_id: body.enterprise_id ?? 1,
      vehicle_type: body.vehicle_type ?? 'shas',
      name: body.name ?? `VEHICLE-${newId}`,
      model_id: body.model_id ?? null,
      model: null,
      serial_number: body.serial_number ?? null,
      registration_number: body.registration_number ?? null,
      status: body.status ?? 'active',
      is_active: body.is_active ?? true,
      active_from: body.active_from ?? now,
      active_to: body.active_to ?? null,
      created_at: now,
      updated_at: now,
    };

    vehiclesDb.push(newVehicle);

    return HttpResponse.json(newVehicle, { status: 201 });
  }),

  // PUT /api/vehicles/:id
  http.put('/api/vehicles/:id', async ({ params, request }) => {
    await delay(mswConfig.delay);

    const id = Number(params.id);
    const vehicleIndex = vehiclesDb.findIndex((v) => v.id === id);

    if (vehicleIndex === -1) {
      return new HttpResponse(null, { status: 404 });
    }

    const body = (await request.json()) as Partial<Vehicle>;
    const updatedVehicle: Vehicle = {
      ...vehiclesDb[vehicleIndex],
      ...body,
      updated_at: new Date().toISOString(),
    };

    vehiclesDb[vehicleIndex] = updatedVehicle;

    return HttpResponse.json(updatedVehicle);
  }),

  // DELETE /api/vehicles/:id
  http.delete('/api/vehicles/:id', async ({ params }) => {
    await delay(mswConfig.delay);

    const id = Number(params.id);
    const vehicleIndex = vehiclesDb.findIndex((v) => v.id === id);

    if (vehicleIndex === -1) {
      return new HttpResponse(null, { status: 404 });
    }

    vehiclesDb.splice(vehicleIndex, 1);

    return new HttpResponse(null, { status: 204 });
  }),
];

File: components/enterprise-service-dev/src/app/schemas/vehicles.py
"""Pydantic schemas для Vehicles."""

from datetime import date, datetime

from pydantic import BaseModel, Field

from app.enums.vehicles import VehicleStatusEnum, VehicleTypeEnum
from app.schemas.common import PaginationBase
from app.schemas.vehicle_models import VehicleModelResponse


class VehicleBase(BaseModel):
    """Базовая схема транспортного средства."""

    enterprise_id: int = Field(..., description="ID предприятия")
    vehicle_type: VehicleTypeEnum = Field(..., description="Тип техники (shas, pdm, vehicle)")
    name: str = Field(..., min_length=1, max_length=100, description="Название техники")
    model_id: int | None = Field(default=None, description="ID модели")
    serial_number: str | None = Field(default=None, max_length=100, description="Серийный номер")
    registration_number: str | None = Field(
        default=None,
        max_length=50,
        description="Регистрационный номер",
    )
    status: VehicleStatusEnum = Field(
        default=VehicleStatusEnum.active,
        description="Статус техники",
    )
    is_active: bool = Field(default=True, description="Техника активна")
    active_from: date | None = Field(default=None, description="Дата начала активности")
    active_to: date | None = Field(default=None, description="Дата окончания активности")


class VehicleCreate(VehicleBase):
    """Схема создания транспортного средства."""


class VehicleUpdate(BaseModel):
    """Схема обновления транспортного средства.

    Все поля optional - передавайте только те, которые нужно изменить.
    """

    enterprise_id: int | None = Field(default=None, description="ID предприятия")
    vehicle_type: VehicleTypeEnum | None = Field(
        default=None,
        description="Тип техники (shas, pdm, vehicle)",
    )
    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
        description="Название техники",
    )
    model_id: int | None = Field(default=None, description="ID модели")
    serial_number: str | None = Field(default=None, max_length=100, description="Серийный номер")
    registration_number: str | None = Field(
        default=None,
        max_length=50,
        description="Регистрационный номер",
    )
    status: VehicleStatusEnum | None = Field(default=None, description="Статус техники")
    is_active: bool | None = Field(default=None, description="Техника активна")
    active_from: date | None = Field(default=None, description="Дата начала активности")
    active_to: date | None = Field(default=None, description="Дата окончания активности")


class VehicleResponse(VehicleBase):
    """Ответ транспортного средства."""

    id: int = Field(..., description="ID техники")
    model: VehicleModelResponse | None = Field(default=None, description="Модель техники")
    created_at: datetime = Field(..., description="Время создания")
    updated_at: datetime = Field(..., description="Время обновления")


class VehicleListResponse(PaginationBase[VehicleResponse]):
    """Ответ списка транспортных средств с пагинацией."""
